Validate tsdb's routing_path

nik9000 opened the pull request:
`routing_path`'s job is to put the same time series on the same node. It
does that by extracting a String directly from the xcontent, hashing it,
and feeding the hash into the shard selection algorithm. I'm happy with
it! But it won't work properly if `routing_path` matches non-dimension
fields or if it matches non-keyword dimensions. This prevents us from
mapping any fields that aren't keyword dimensions that "line up" with
`routing_path`.

Let's talk about why `routing_path` can't do it's job if it matches any
non-dimension fields. Well, imagine `routing_path` is `[dim, foo]` and
only `dim` is a dimension. It'll *still* hash `foo`'s values into the
routing key. So, say you get documents like:
```
{"dim": "a", "foo": "1"}
{"dim": "a", "foo": "1"}
{"dim": "a", "foo": "2"}
```

The third document could be routed to a different shard than the first
two! Which would be a disaster because it'd cut the time series
identified by `"dim":"a"` into pieces!

Now let's talk about when `routing_path` matches a non-keyword
dimension. Imagine `routing_path` is `[kwd, int]` and we send these
documents:
```
{"kwd": "a", "int": "1"}
{"kwd": "a", "int": "01"}
```

Both of these documents belong to the time series identified by
`"dim":"a","int":1` but the `routing_path` code just reads strings so
it'll route them into separate shards. Also bad! Also forbidden by this
change.

nik9000 committed:
Validate tsdb's routing_path

`routing_path`'s job is to put the same time series on the same node. It
does that by extracting a String directly from the xcontent, hashing it,
and feeding the hash into the shard selection algorithm. I'm happy with
it! But it won't work properly if `routing_path` matches non-dimension
fields or if it matches non-keyword dimensions. This prevents us from
mapping any fields that aren't keyword dimensions that "line up" with
`routing_path`.

Let's talk about why `routing_path` can't do it's job if it matches any
non-dimension fields. Well, imagine `routing_path` is `[dim, foo]` and
only `dim` is a dimension. It'll *still* hash `foo`'s values into the
routing key. So, say you get documents like:
```
{"dim": "a", "foo": "1"}
{"dim": "a", "foo": "1"}
{"dim": "a", "foo": "2"}
```

The third document could be routed to a different shard than the first
two! Which would be a disaster because it'd cut the time series
identified by `"dim":"a"` into pieces!

Now let's talk about when `routing_path` matches a non-keyword
dimension. Imagine `routing_path` is `[kwd, int]` and we send these
documents:
```
{"kwd": "a", "int": "1"}
{"kwd": "a", "int": "01"}
```

Both of these documents belong to the time series identified by
`"dim":"a","int":1` but the `routing_path` code just reads strings so
it'll route them into separate shards. Also bad! Also forbidden by this
change.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 committed:
Merge branch 'master' into index_routing_from_source_validate_dims

imotov reviewed (approved):
LGTM in general, added one suggestion.

nik9000 committed:
Move routing calculation

This slightly moves the routing calculation out of a difficult to reason
about `switch` statement and into real OO method implementation. Its a
tiny tiny change but it makes me feel much better about it.

nik9000 committed:
Revert "Move routing calculation"

This reverts commit 9a040d421bf17a0731054a7d188514f285bd4260.

nik9000 committed:
privacy

nik9000 committed:
Handle warnings

nik9000 committed:
Merge branch 'master' into index_routing_from_source_validate_dims
